lib/plugins/__init__.py

```diff
@@ -0,0 +1,61 @@
+from os.path import expanduser
+from lib import configuration
+import logging
+log = logging.getLogger('fabric.fabalicious.plugins')
+
+def init(root_folder, settings_name, folder_name, plugin_name, categories_filter):
+  from yapsy.PluginManager import PluginManager
+
+  plugin_dirs = []
+  if configuration.getSettings(settings_name):
+    plugin_dirs.append(configuration.fabfile_basedir + '/' + configuration.getSettings(settings_name))
+
+  plugin_dirs.append(configuration.fabfile_basedir + '/.fabalicious/' + folder_name)
+  plugin_dirs.append(configuration.fabfile_basedir + '/plugins/' + folder_name)
+
+  plugin_dirs.append(expanduser("~") + '/.fabalicious/plugins/' + folder_name)
+  plugin_dirs.append(root_folder + '/plugins/' + folder_name)
+  log.debug("Looking for %s-plugins in %s" % (plugin_name, ", ".join(plugin_dirs)))
+
+  manager = PluginManager()
+  manager.setPluginPlaces(plugin_dirs)
+  manager.setCategoriesFilter(categories_filter)
+  manager.collectPlugins()
+
+  # Activate all loaded plugins
+  for pluginInfo in manager.getAllPlugins():
+    manager.activatePluginByName(pluginInfo.name)
+
+  result = {}
+  for plugin in manager.getAllPlugins():
+
+    if hasattr(plugin.plugin_object, 'aliases') and isinstance(plugin.plugin_object.aliases, list):
+      for alias in plugin.plugin_object.aliases:
+        result[alias] = plugin.plugin_object
+    else:
+      result[plugin.name] = plugin.plugin_object
+  return result
+
+
+def getTasks(root_folder):
+  try:
+    __import__('imp').find_module('yapsy')
+    from task import ITaskPlugin
+
+    return init(root_folder, 'customTasksFolder', 'tasks', 'task', { "Task": ITaskPlugin })
```

Comment:
NOTE: Just a discussion, not asking you to implement it.

yapsy searches into sub-folder too, so do we need to specify Task specific folder? 

The advantage of current solution is some time saved during plugin discovery.

The tasks will be filtered based on the class it extends. So this gives a good degree of freedom on the folder structure that developer could follow to organise his plugins.

Possible solution
1. use `customPluginsFolder` and add subfolders `task` and `methods` inside to keep it organised.
2. use `customPluginsFolders` (notice the `s`), and accept an array of locations that should be searched for.
3. In normal use case, the developer writing plugins would have to add the entry in fabfile.yaml just once.

Thinking about it further, lets  alter the key a little more and make it,

```
customPlugins:
  folder: ".fabalicious/plugins"
```
We might want to introduce/capture few more parameters in the future.